crates/wadm-types/src/validation.rs

```diff
@@ -722,6 +723,50 @@ pub fn is_valid_label_name(name: &str) -> bool {
             .all(|c| c.is_ascii_alphanumeric() || c == '-' || c == '_' || c == '.')
 }
 
+/// Checks whether a manifest contains "duplicate" links.
+///
+/// Multiple links from the same source with the same name, namespace, package
+/// and interface are considered duplicate links.
+fn check_duplicate_links(manifest: &Manifest) -> Vec<ValidationFailure> {
+    let mut failures = Vec::new();
+    let mut link_ids = HashSet::new();
+    for component in manifest.components() {
+        if let Some(traits) = component.traits.as_ref() {
+            link_ids.clear();
+            let link_traits = traits.iter().filter(|t| t.is_link());
+            for link_trait in link_traits {
+                if let TraitProperty::Link(LinkProperty {
+                    name,
+                    namespace,
+                    package,
+                    interfaces,
+                    ..
+                }) = &link_trait.properties
+                {
+                    for interface in interfaces {
+                        if !link_ids.insert((
+                            name.clone()
+                                .unwrap_or_else(|| DEFAULT_LINK_NAME.to_string()),
+                            namespace,
+                            package,
+                            interface,
+                        )) {
+                            failures.push(ValidationFailure::new(
+                                ValidationFailureLevel::Error,
+                                format!(
+                                    "Duplicate links found inside component: {}",
```

Comment:
I wonder if the error message should contain information about which link name (or perhaps namespace/package/interface triplet) conflicted so that it might be easier to track down?

Reply:
I guess both the link name and the namespace/package can be included in the message?